src/nf/nf_layer_submodule.f90

```diff
@@ -269,6 +269,91 @@ impure elemental module subroutine print_info(self)
   end subroutine print_info
 
 
+  pure module function get_num_params(self) result(num_params)
+    implicit none
+    class(layer), intent(in) :: self
+    integer :: num_params
+
+    select type (this_layer => self % p)
+      type is (input1d_layer)
+        num_params = 0
+      type is (input3d_layer)
+        num_params = 0
+      type is (dense_layer)
+        num_params = this_layer % get_num_params()
+      type is (conv2d_layer)
+        num_params = this_layer % get_num_params()
+      type is (maxpool2d_layer)
+        num_params = 0
+      type is (flatten_layer)
+        num_params = 0
+      type is (reshape3d_layer)
+        num_params = 0
+      class default
+        error stop 'Unknown layer type.'
+    end select
+
+  end function get_num_params
+
+
+  pure module subroutine get_params(self, params)
+    class(layer), intent(in) :: self
+    real, allocatable, intent(in out) :: params(:)
+
+    select type (this_layer => self % p)
+      type is (input1d_layer)
+         ! No parameters to get.
+      type is (input3d_layer)
+         ! No parameters to get.
+      type is (dense_layer)
+        call this_layer % get_params(params)
+      type is (conv2d_layer)
+        call this_layer % get_params(params)
+      type is (maxpool2d_layer)
+        ! No parameters to get.
+      type is (flatten_layer)
+        ! No parameters to get.
+      type is (reshape3d_layer)
+        ! No parameters to get.
+      class default
+        error stop 'Unknown layer type.'
+    end select
+
+  end subroutine get_params
+
+
+  impure module function set_params(self, params) result(consumed)
```

Comment:
@jvo203 Why make `set_params` a function here? A subroutine seems more natural for a method that sets the values in an object.

Reply:
There are two equivalent ways:

1. set_params is a function that returns the number of parameters "consumed" by this particular layer (in other words moves the array cursor further ahead)

or

2. set_params is a subroutine where the consumed argument is "inout", the number of consumed array entries will be added to the consumed argument, to be used by other layers further down the line